### PR TITLE
Adding PiP button from video control overlay to video tab bar

### PR DESCRIPTION
**Pull request by BandarHL**

### CODE REVIEW

**Comment by PoomSmart** [comment]
Thanks for the PR. I want to raise some concerns.

- Do you believe users prefer PiP button at the video tab bar over the video control overlay?
- Do you agree to let users choose between it being in the tab bar and the overlay?
- There is a potential conflict with YouTube tweaks that manage the buttons at the video tab bar (uYou for example) that the hardcoded index 4 might be used against the array of total length less than 5.

**Comment by BandarHL** [comment]
Yeah I believe is better because it is closer to the user hand, and for another reason the view is scrollable, so the user can benefit from it, for example, if the used is in a situation unable to extend his finger to the top of the screen,  just drags the view and press the button in an easier way (I hope you get the idea). Or we just add new option that let user select what prefer.
And for uYou I never used before and its true, uYou take the index 4 im trying to fix the issue.

**Comment by PoomSmart** [comment]
@BandarHL If a video is in full-screen mode I still see the need of having PiP button on the video control overlay. That said, adding PiP button is preferred over moving.

File: Tweak.x [hunk 1]
```diff
@@ -112,101 +115,91 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
-%hook YTMainAppVideoPlayerOverlayViewController
-
-- (void)updateTopRightButtonAvailability {
-    %orig;
-    YTMainAppVideoPlayerOverlayView *v = [self videoPlayerOverlayView];
-    YTMainAppControlsOverlayView *c = [v valueForKey:@"_controlsOverlayView"];
-    c.pipButton.hidden = !UsePiPButton();
-    [c setNeedsLayout];
+%hook YTIIcon
+- (id)iconImageWithColor:(UIColor *)arg1 {
```
Comment (PoomSmart) [inline Tweak.x:119]:
```suggestion

- (UIImage *)iconImageWithColor:(UIColor *)color {
```

File: Tweak.x [hunk 2]
```diff
@@ -103,7 +106,7 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
     @try {
         hotConfig = [self valueForKey:@"_hotConfig"];
     } @catch (id ex) {
-        hotConfig = [[self gimme] instanceForType:%c(YTHotConfig)];
+ //       hotConfig = [[self gimme] instanceForType:%c(YTHotConfig)];
```
[the remark this answers is not in the document]
Comment (BandarHL) [reply Tweak.x:109]: YTPlayerViewController doesn't have 'gimme' instance how its compiled with you?

**Comment by BandarHL** [comment]
but there is something weird, why uYou keep blocking me from add objects to 'SupportedRenderers' array?

**Comment by BandarHL** [comment]
ahhhhhhh I wanna kill myself, hours of working and the solution is fucking simple. I just pass the '%orig' at the end of 'createActionViewsFromSupportedRenderers' method.

File: Tweak.x (hunk 2 above)
[the remark this answers is not in the document]
Comment (PoomSmart) [reply Tweak.x:109]: I might have accidentally removed it from the header.. :/

**Comment by BandarHL** [comment]
- i add new option for it. what you think about it?

**Comment by PoomSmart** [comment]
@BandarHL Looking good. As for the functionality, I would say if either PiP button on video control overlay or video tab bar is enabled, the default PiP behavior (dismiss to PiP) should be disabled.

File: Tweak.x [hunk 3]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+        if ([image respondsToSelector:@selector(imageFlippedForRightToLeftLayoutDirection)])
+            image = [image imageFlippedForRightToLeftLayoutDirection];
+        return image;
+    } else {
+        return %orig;
+    }
+}
+%end
+
+%hook YTSlimVideoScrollableDetailsActionsView
```
Comment (PoomSmart) [inline Tweak.x:144]:
```suggestion
%hook YTSlimVideoScrollableDetailsActionsView

```

File: Tweak.x [hunk 4]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+- (void)createActionViewsFromSupportedRenderers:(NSMutableArray *)renderers { // for old YouTube version
+    if (UseTabBarPiPButton()) {
+        YTISlimMetadataButtonSupportedRenderers *PiPButton = [self makeNewButtonWithTitle:@"PiP" iconType:007 BrowseId:@"YouPiP.pip.command"];
+        if (![renderers containsObject:PiPButton]) {
+            [renderers addObject:PiPButton];
+        }
+    }
+    %orig;
+}
```
Comment (PoomSmart) [inline Tweak.x:153]:
```suggestion
}

```

File: Tweak.x [hunk 5]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+- (void)createActionViewsFromSupportedRenderers:(NSMutableArray *)renderers withElementsContextBlock:(id)arg2 {
+    if (UseTabBarPiPButton()) {
+        YTISlimMetadataButtonSupportedRenderers *PiPButton = [self makeNewButtonWithTitle:@"PiP" iconType:007 BrowseId:@"YouPiP.pip.command"];
+        if (![renderers containsObject:PiPButton]) {
+            [renderers addObject:PiPButton];
+        }
+    }
+    %orig;
+}
+%new - (YTISlimMetadataButtonSupportedRenderers *)makeNewButtonWithTitle:(NSString *)title iconType:(int)IconType BrowseId:(NSString *)browseid {
```
Comment (PoomSmart) [inline Tweak.x:163]:
```suggestion
%new
- (YTISlimMetadataButtonSupportedRenderers *)makeNewButtonWithTitle:(NSString *)title iconType:(int)iconType browseId:(NSString *)browseId {
```

File: Tweak.x [hunk 6]
```diff
@@ -15,6 +15,22 @@
...
 #import "../YouTubeHeader/QTMIcon.h"
+#import "../YouTubeHeader/YTWatchController.h"
+#import "../YouTubeHeader/YTNGWatchController.h"
+#import "../YouTubeHeader/YTSlimVideoScrollableActionBarCellController.h"
+#import "../YouTubeHeader/YTSlimVideoScrollableDetailsActionsView.h"
+#import "../YouTubeHeader/YTSlimVideoDetailsActionView.h"
+#import "../YouTubeHeader/YTPageStyleController.h"
+
+@interface YTSlimVideoScrollableDetailsActionsView ()
+- (YTISlimMetadataButtonSupportedRenderers *)makeNewButtonWithTitle:(NSString *)title iconType:(int)IconType BrowseId:(NSString *)browseid;
```
Comment (PoomSmart) [inline Tweak.x:26]:
```suggestion
- (YTISlimMetadataButtonSupportedRenderers *)makeNewButtonWithTitle:(NSString *)title iconType:(int)iconType browseId:(NSString *)browseId;
```

File: Tweak.x [hunk 7]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+    if (UseTabBarPiPButton()) {
+        YTISlimMetadataButtonSupportedRenderers *PiPButton = [self makeNewButtonWithTitle:@"PiP" iconType:007 BrowseId:@"YouPiP.pip.command"];
+        if (![renderers containsObject:PiPButton]) {
+            [renderers addObject:PiPButton];
+        }
+    }
+    %orig;
+}
+%new - (YTISlimMetadataButtonSupportedRenderers *)makeNewButtonWithTitle:(NSString *)title iconType:(int)IconType BrowseId:(NSString *)browseid {
+    YTISlimMetadataButtonSupportedRenderers *SupportedRenderer = [[%c(YTISlimMetadataButtonSupportedRenderers) alloc] init];
```
Comment (PoomSmart) [inline Tweak.x:164]:
```suggestion
    YTISlimMetadataButtonSupportedRenderers *supportedRenderer = [[%c(YTISlimMetadataButtonSupportedRenderers) alloc] init];
```

File: Tweak.x [hunk 8]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+    YTISlimMetadataButtonRenderer *MetadataButtonRenderer = [[%c(YTISlimMetadataButtonRenderer) alloc] init];
```
Comment (PoomSmart) [inline Tweak.x:165]:
```suggestion
    YTISlimMetadataButtonRenderer *metadataButtonRenderer = [[%c(YTISlimMetadataButtonRenderer) alloc] init];
```
Under normal circumstances should the variables be named non-capitalized.

File: Tweak.x [hunk 9]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+    YTIButtonRenderer *button = [[%c(YTIButtonRenderer) alloc] init];
+    YTIIcon *icon = [[%c(YTIIcon) alloc] init];
+
+    [endPoint setBrowseId:browseid];
+    [command setBrowseEndpoint:endPoint];
+    [icon setIconType:IconType];
+    
+    [button setStyle:8]; // Opacity style
+    [button setTooltip:title];
+    [button setSize:1]; // Defualt size
```
Comment (PoomSmart) [inline Tweak.x:178]:
```suggestion
    [button setSize:1]; // Default size
```

File: Tweak.x [hunk 10]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+%new - (YTISlimMetadataButtonSupportedRenderers *)makeNewButtonWithTitle:(NSString *)title iconType:(int)IconType BrowseId:(NSString *)browseid {
+    YTISlimMetadataButtonSupportedRenderers *SupportedRenderer = [[%c(YTISlimMetadataButtonSupportedRenderers) alloc] init];
+    YTISlimMetadataButtonRenderer *MetadataButtonRenderer = [[%c(YTISlimMetadataButtonRenderer) alloc] init];
+    YTIButtonSupportedRenderers *ButtonSupportedRenderer = [[%c(YTIButtonSupportedRenderers) alloc] init];
+    YTIBrowseEndpoint *endPoint = [[%c(YTIBrowseEndpoint) alloc] init];
+    YTICommand *command = [[%c(YTICommand) alloc] init];
+    YTIButtonRenderer *button = [[%c(YTIButtonRenderer) alloc] init];
+    YTIIcon *icon = [[%c(YTIIcon) alloc] init];
+
+    [endPoint setBrowseId:browseid];
```
Comment (PoomSmart) [inline Tweak.x:172]: If this is an objc property:
```suggestion
    endPoint.browseId = browseId;
```

File: Tweak.x [hunk 11]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+    [button setText:[%c(YTIFormattedString) formattedStringWithString:title]];
+    [button setIcon:icon];
+    [button setNavigationEndpoint:command];
+    
+    [ButtonSupportedRenderer setButtonRenderer:button];
+    [MetadataButtonRenderer setButton:ButtonSupportedRenderer];
+    [SupportedRenderer setSlimMetadataButtonRenderer:MetadataButtonRenderer];
+    
+    return SupportedRenderer;
+}
```
Comment (PoomSmart) [inline Tweak.x:189]:
```suggestion
}

```

File: Tweak.x [hunk 12]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+%end
+
+%hook YTSlimVideoDetailsActionView
```
Comment (PoomSmart) [inline Tweak.x:192]:
```suggestion
%hook YTSlimVideoDetailsActionView

```

File: Tweak.x [hunk 13]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+                YTNGWatchController *NGWMetadataPanelStateProvider = [_delegate valueForKey:@"_ngwMetadataPanelStateProvider"];
+                if ([[NGWMetadataPanelStateProvider valueForKey:@"_playerViewController"] isKindOfClass:%c(YTPlayerViewController)]) {
+                    YTPlayerViewController *playerViewController = [NGWMetadataPanelStateProvider valueForKey:@"_playerViewController"];
+                    FromUser = YES;
+                    bootstrapPiP(playerViewController, YES);
+                }
+            }
+        }
+    } else {
+        return %orig;
```
Comment (PoomSmart) [inline Tweak.x:216]:
```suggestion
        %orig;
```

File: Tweak.x [hunk 14]
```diff
@@ -110,7 +124,101 @@ static void bootstrapPiP(YTPlayerViewController *self, BOOL playPiP) {
...
+    }
+}
```
Comment (PoomSmart) [inline Tweak.x:218]:
```suggestion
}

```

File: Tweak.x [hunk 15]
```diff
@@ -209,6 +317,7 @@ static NSMutableArray *topControls(YTMainAppControlsOverlayView *self, NSMutable
 
 %end
 
+
```
Comment (PoomSmart) [inline Tweak.x:320]:
```suggestion
```

**Review by PoomSmart — CHANGES_REQUESTED** [review]

**Review by PoomSmart — APPROVED** [review]

**Pull request merged**